tests/bsim/bluetooth/host/att/timeout/main.c

```diff
@@ -0,0 +1,288 @@
+
+static ssize_t read_chrc(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			 uint16_t buf_len, uint16_t offset)
+{
+	ssize_t read_len;
+
+	LOG_INF("ATT timeout will %sbe triggered", trigger_att_timeout ? "" : "not ");
+
+	if (trigger_att_timeout) {
+		/* Sleep longer than ATT Timeout (section 3.3.3). */
+		k_sleep(K_SECONDS(BT_ATT_TIMEOUT_SEC + 1));
+	}
+
+	__ASSERT_NO_MSG(offset == 0);
+	read_len = buf_len;
+
+	__ASSERT_NO_MSG(read_len >= 2);
+	sys_put_le16(read_len, buf);
+
+	return read_len;
+}
+
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(UUID_1),
+	BT_GATT_CHARACTERISTIC(UUID_2, BT_GATT_CHRC_READ, BT_GATT_PERM_READ, read_chrc, NULL, NULL),
+};
+
+static struct bt_gatt_service svc = {
+	.attrs = attrs,
+	.attr_count = ARRAY_SIZE(attrs),
+};
+
+static void bs_sync_all_log(char *log_msg)
+{
+	/* Everyone meets here. */
+	bt_testlib_bs_sync_all();
+
+	if (get_device_nbr() == 0) {
+		LOG_WRN("Sync point: %s", log_msg);
+	}
+
+	/* Everyone waits for d0 to finish logging. */
+	bt_testlib_bs_sync_all();
+}
+
+static inline void bt_enable_quiet(void)
+{
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_ERR);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_ERR);
+
+	EXPECT_ZERO(bt_enable(NULL));
+
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_INF);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_INF);
+}
+
+static void peripheral_setup(enum bt_att_chan_opt bearer, bool timeout)
+{
+	struct bt_conn *conn = NULL;
+
+	EXPECT_ZERO(bt_testlib_adv_conn(&conn, BT_ID_DEFAULT, bt_get_name()));
+
+	trigger_att_timeout = timeout;
+}
+
+static struct bt_conn *central_setup(enum bt_att_chan_opt bearer, bool timeout)
+{
+	bt_addr_le_t adva;
+	struct bt_conn *conn = NULL;
+
+	EXPECT_ZERO(bt_testlib_scan_find_name(&adva, "peripheral"));
+	EXPECT_ZERO(bt_testlib_connect(&adva, &conn));
+
+	/* Establish EATT bearers. */
+	EXPECT_ZERO(bt_testlib_secure(conn, BT_SECURITY_L2));
+
+	while (bt_eatt_count(conn) == 0) {
+		k_msleep(100);
+	};
+
+	return conn;
+}
+
+static void central_read(struct bt_conn *conn, enum bt_att_chan_opt bearer, bool timeout)
+{
+	uint16_t actual_read_len;
+	uint16_t remote_read_send_len;
+	uint16_t handle = 0;
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(attr_value, sizeof(remote_read_send_len));
+
+	err = bt_testlib_att_read_by_type_sync(&attr_value, &actual_read_len, &handle, NULL, conn,
+					       bearer, UUID_2, BT_ATT_FIRST_ATTRIBUTE_HANDLE,
+					       BT_ATT_LAST_ATTRIBUTE_HANDLE);
+
+	if (timeout) {
+		__ASSERT(err == BT_ATT_ERR_UNLIKELY, "Unexpected error %d", err);
+	} else {
+		__ASSERT(!err, "Unexpected error %d", err);
+		__ASSERT(attr_value.len >= sizeof(remote_read_send_len),
+			 "Remote sent too little data.");
+		remote_read_send_len = net_buf_simple_pull_le16(&attr_value);
+		__ASSERT(remote_read_send_len == actual_read_len, "Length mismatch. %u %u",
+			 remote_read_send_len, actual_read_len);
+	}
+}
+
+static void wait_for_disconnect(bool timeout)
```

Comment:
I suggest we inline this function for clarity. It's very core and specific to the test.